test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java

```diff
@@ -0,0 +1,431 @@
+                                                                 (partitioner, tokens) -> new VersionedValue.VersionedValueFactory(partitioner).normal(tokens),
+                                                                 peer.config().getString("partitioner"),
+                                                                 peer.config().getString("initial_token")),
+                                            unsafeVersionedToken(target,
+                                                                 ApplicationState.STATUS_WITH_PORT,
+                                                                 (partitioner, tokens) -> new VersionedValue.VersionedValueFactory(partitioner).normal(tokens),
+                                                                 peer.config().getString("partitioner"),
+                                                                 peer.config().getString("initial_token")),
+                                            unsafeReleaseVersion(target,
+                                                                 peer.config().getString("partitioner"),
+                                                                 peer.getReleaseVersionString())));
+        };
+    }
+
+    public static InstanceAction statusToLeaving(IInvokableInstance newNode)
+    {
+        return (instance) -> {
+            changeGossipState(instance,
+                              newNode,
+                              Arrays.asList(tokens(newNode),
+                                            statusLeaving(newNode),
+                                            statusWithPortLeaving(newNode)));
+        };
+    }
+
+    public static InstanceAction bootstrap()
+    {
+        return new BootstrapAction();
+    }
+
+    public static InstanceAction bootstrap(boolean joinRing, Duration waitForBootstrap, Duration waitForSchema)
+    {
+        return new BootstrapAction(joinRing, waitForBootstrap, waitForSchema);
+    }
+
+    public static InstanceAction disseminateGossipState(IInvokableInstance newNode)
+    {
+        return new DisseminateGossipState(newNode);
+    }
+
+    public static InstanceAction pullSchemaFrom(IInvokableInstance pullFrom)
+    {
+        return new PullSchemaFrom(pullFrom);
+    }
+
+    private static InstanceAction disableBinary()
+    {
+        return (instance) -> {
+            instance.runOnInstance(StorageService.instance::stopNativeTransport);
+        };
+    }
+
+    private static class DisseminateGossipState implements InstanceAction
+    {
+        final Map<InetSocketAddress, byte[]> gossipState;
+
+        public DisseminateGossipState(IInvokableInstance... from)
+        {
+            gossipState = new HashMap<>();
+            for (IInvokableInstance node : from)
+            {
+                byte[] epBytes = node.callsOnInstance(() -> {
+                    EndpointState epState = Gossiper.instance.getEndpointStateForEndpoint(FBUtilities.getBroadcastAddressAndPort());
```

Comment:
Just to make sure, I've switched to 

```
            cluster.run(statusToBootstrap(newInstance), newInstance.config().num());
            cluster.run(disseminateGossipState(newInstance));
```

instead of just `statusToBootstrap`, and everything seems to still work. Since each node is calling enquiring its own endpoint, this seems to be fine at first glance, but if you elaborate on where this breaks, I can take another look.

Reply:
I think I found what I was thinking about `org.apache.cassandra.service.StorageService#loadRingState`

```
private void loadRingState()
    {
        if (Boolean.parseBoolean(System.getProperty("cassandra.load_ring_state", "true")))
        {
            logger.info("Loading persisted ring state");
            Multimap<InetAddressAndPort, Token> loadedTokens = SystemKeyspace.loadTokens();
            Map<InetAddressAndPort, UUID> loadedHostIds = SystemKeyspace.loadHostIds();
            for (InetAddressAndPort ep : loadedTokens.keySet())
            {
                if (ep.equals(FBUtilities.getBroadcastAddressAndPort()))
                {
                    // entry has been mistakenly added, delete it
                    SystemKeyspace.removeEndpoint(ep);
                }
                else
                {
                    if (loadedHostIds.containsKey(ep))
                        tokenMetadata.updateHostId(loadedHostIds.get(ep), ep);
                    Gossiper.runInGossipStageBlocking(() -> Gossiper.instance.addSavedEndpoint(ep, loadedHostIds.get(ep), loadedTokens.get(ep)));
                }
            }
        }
    }
```

We remove from peers table, so think I was wrong here.